ambry-store/src/main/java/com.github.ambry.store/IndexValue.java

```diff
@@ -13,110 +13,343 @@
  */
 package com.github.ambry.store;
 
+import com.github.ambry.utils.Time;
 import com.github.ambry.utils.Utils;
 import java.nio.ByteBuffer;
 
 
 /**
  * Represents the blob value stored in the index for a key.
+ *
+ * Version_0
+ * - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+ * | Blob Size |   Offset  |  Flags  | Expiration Time | Orig msg  |
+ * | (8 bytes) | (8 bytes) | (1 byte)|     in Ms       | offset    |
+ * |           |           |         |     ( 8 bytes)  | (8 bytes) |
+ * - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+ *
+ * Version_1
+ * - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+ * | Blob Size |   Offset  |  Flags  | Expiration Time | Orig msg  | OperationTime | ServiceId | ContainerId |
+ * | (8 bytes) | (8 bytes) | (1 byte)|   in  Secs      | offset    |  (4 bytes)    | (2 bytes) | (2 bytes)   |
+ * |           |           |         |     ( 4 bytes)  | (8 bytes) |               |           |             |
+ * - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+ *
  */
-
 class IndexValue {
   enum Flags {
     Delete_Index
   }
 
-  private static int Blob_Size_In_Bytes = 8;
-  private static int Offset_Size_In_Bytes = 8;
-  private static int Flag_Size_In_Bytes = 1;
-  private static int Expires_At_Ms_Size_In_Bytes = 8;
-  private static int Original_Message_Offset_Size_In_Bytes = 8;
+  private final static int BLOB_SIZE_IN_BYTES = 8;
+  private final static int OFFSET_SIZE_IN_BYTES = 8;
+  private final static int FLAG_SIZE_IN_BYTES = 1;
+  private final static int EXPIRES_AT_MS_SIZE_IN_BYTES_V0 = 8;
+  private final static int ORIGINAL_MESSAGE_OFFSET_SIZE_IN_BYTES = 8;
+
+  private final static int EXPIRES_AT_SECS_SIZE_IN_BYTES_V1 = 4;
+  private final static int OPERATION_TIME_SECS_SIZE_IN_BYTES_V1 = 4;
+  private final static int SERVICE_ID_SIZE_IN_BYTES = 2;
+  private final static int CONTAINER_ID_SIZE_IN_BYTES = 2;
+
+  // relative offsets for version 0
+  private final static int OFFSET_RELATIVE_OFFSET = BLOB_SIZE_IN_BYTES;
```

Comment:
why not store the actual values as class variables instead of trying to calculate offsets and reading correctly in the `ByteBuffer`? This looks a whole lot more complicated than just maintaining class variables and generating the right `ByteBuffer` on a `toBytes()` call.